packages/nextjs/src/server/clerkClient.ts

```diff
@@ -27,6 +33,63 @@ const clerkClient = createClerkClient({
     disabled: TELEMETRY_DISABLED,
     debug: TELEMETRY_DEBUG,
   },
+};
+
+export const clerkClientStorage = new AsyncLocalStorage<ClerkRequestData>();
+
+const createClerkClientWithOptions: typeof createClerkClient = options =>
+  createClerkClient({ ...clerkClientDefaultOptions, ...options });
+
+const clerkClientSingleton = createClerkClient(clerkClientDefaultOptions);
+
+/**
+ * @deprecated
+ * This singleton is deprecated and will be removed in a future release. Please use `clerkClient()` as a function instead.
+ */
+const clerkClientSingletonProxy = new Proxy(clerkClientSingleton, {
+  get(target, prop, receiver) {
+    deprecated('clerkClient object', 'Use `clerkClient()` as a function instead.');
+
+    return Reflect.get(target, prop, receiver);
+  },
 });
 
+/**
+ * Constructs a BAPI client that accesses request data within the runtime.
+ * Necessary if middleware dynamic keys are used.
+ */
+const clerkClientForRequest = () => {
+  let requestData: ClerkRequestData | undefined;
+
+  /**
+   * For BAPI client usage inside middleware runtime, fallbacks to AsyncLocalStorage to access request data
+   */
+  const clerkClientStore = clerkClientStorage.getStore();
+  if (clerkClientStore) {
+    requestData = clerkClientStore;
+  } else {
+    /**
+     * For BAPI usage from application server, fallbacks to access request data via `NextRequest`
+     */
+    const request = buildRequestLike();
+    const encryptedRequestData = getHeader(request, constants.Headers.ClerkRequestData);
+    requestData = decryptClerkRequestData(encryptedRequestData);
+  }
+
+  if (requestData?.secretKey || requestData?.publishableKey) {
+    return createClerkClientWithOptions({
+      secretKey: requestData.secretKey,
+      publishableKey: requestData.publishableKey,
+    });
+  }
+
+  return clerkClient;
+};
+
+const clerkClient: ClerkClient & typeof clerkClientForRequest = Object.assign(
```

Comment:
Must update docs to call `clerkClient()` instead of `clerkClient` https://clerk.com/docs/references/nextjs/route-handlers#interact-with-clerks-backend-api